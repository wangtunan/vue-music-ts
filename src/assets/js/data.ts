import { TabConfig } from '@/types/index'

export const tabList: TabConfig[] = [
  { name: '推荐', path: '/recommend' },
  { name: '歌手', path: '/singer' },
  { name: '排行', path: '/rank' },
  { name: '搜索', path: '/search' }
]
