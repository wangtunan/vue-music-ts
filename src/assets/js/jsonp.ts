import originJsonp from 'jsonp'
import { Jsonpoptions } from '@/types/index'
import { combineParams } from '@/utils/utils'

export default function jsonp (url: string, data: any, options: Jsonpoptions): Promise<any> {
  url += (url.indexOf('?') < 0 ? '?' : '&') + combineParams(data)
  return new Promise((resolve, reject) => {
    originJsonp(url, options, (err, data) => {
      if (!err) {
        resolve(data)
      } else {
        reject(err)
      }
    })
  })
}
