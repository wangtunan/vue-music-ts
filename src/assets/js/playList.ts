import Song from './song'
import { Component, Vue, Watch } from 'vue-property-decorator'
import { Getter } from 'vuex-class'
@Component
export default class PlayList extends Vue {
  @Getter('playList') playList!: Song[]
  @Watch('playList')
  onPlayListChange () {
    this.handlePlayList()
  }
  public handlePlayList () {
    throw new Error('component must implement handlePlayList method')
  }
}
