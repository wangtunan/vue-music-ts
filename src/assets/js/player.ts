import Song from '@/assets/js/song'
import { Component, Vue } from 'vue-property-decorator'
import { Getter, Mutation, Action } from 'vuex-class'
import { PlayMode } from '@/types/player'
@Component
export default class Player extends Vue {
  @Getter('mode') mode!: number
  @Getter('currentSong') currentSong!: Song
  @Getter('favoriteList') favoriteList!: Song[]
  @Mutation('player/SET_PLAY_MODE') setPlayMode!: (mode: number) => void
  @Mutation('player/SET_PLAY_STATE') setPlayState!: (playing: boolean) => void
  @Action('player/saveFavoriteList') saveFavoriteList!: (song: Song) => void
  @Action('player/deleteFavoriteList') deleteFavoriteList!: (song: Song) => void

  // methods方法
  public handleModeChange () {
    const mode = (this.mode + 1) % 3
    this.setPlayMode(mode)
  }
  public handleToggleFavorite () {
    if (this.isFavorite(this.currentSong)) {
      this.deleteFavoriteList(this.currentSong)
    } else {
      this.saveFavoriteList(this.currentSong)
    }
  }
  private isFavorite (song: Song): boolean {
    return this.favoriteList.findIndex(item => {
      return item.id === song.id
    }) > -1
  }
  private getFavoriteIcon (song: Song): string {
    return this.isFavorite(song) ? 'icon-favorite' : 'icon-not-favorite'
  }

  // 计算属性
  private get modeIcon () {
    return this.mode === PlayMode.sequence ? 'icon-sequence' : this.mode === PlayMode.loop ? 'icon-loop' : 'icon-random'
  }
  private get faviroteIcon () {
    return this.getFavoriteIcon(this.currentSong)
  }
}
