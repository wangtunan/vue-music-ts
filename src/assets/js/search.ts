import { Component, Vue } from 'vue-property-decorator'
@Component
export default class Search extends Vue {
  protected keyword = ''
  protected searchHistory: string[] = ['曹操', '林俊杰', '周杰伦', '王菲', '潘玮柏', 'S.H.E', '张宇', '马天宇', '陈坤', '五月天', '伍佰']

  // methods方法
  public handleSearch (keyword: string): void {
    this.keyword = keyword.trim()
  }
  public handleClearClick (): void {
    // TDD
  }
  public handleDeleteClick (): void {
    // TDD
  }
}
