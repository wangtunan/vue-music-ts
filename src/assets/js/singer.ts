export default class Singer {
  id: string
  name: string
  avatar: string
  constructor (id: string, name: string) {
    this.id = id
    this.name = name
    this.avatar = `https://y.gtimg.cn/music/photo_new/T001R300x300M000${id}.jpg?max_age=2592000`
  }
}
