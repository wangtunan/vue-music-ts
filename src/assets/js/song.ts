import { SongData } from '@/types/search'
import { getSongUrl, getLyric } from '@/api/song'
import { SongUrlMap } from '@/types/song'
import { ERR_OK } from '@/api/config'
import { Base64 } from 'js-base64'
function filterSinger (singer: [] | any): string {
  const result: string[] = []
  if (!singer) {
    return ''
  }
  if (Array.isArray(singer)) {
    singer.forEach(item => {
      result.push(item.name)
    })
  }
  return result.join('/')
}

interface ISong {
  id: string;
  mid: string;
  singer: string;
  name: string;
  album: string;
  duration: number;
  image: string;
  filename: string;
  url: string;
  type: string;
  lyric: string;
}
export default class Song implements ISong {
  public id: string;
  public mid: string;
  public singer: string;
  public name: string;
  public album: string;
  public duration: number;
  public image: string;
  public filename: string;
  public url: string;
  public type: string;
  public lyric: string;
  constructor ({ id, mid, singer, name, album, duration, image, filename, url, type, lyric }: ISong) {
    this.id = id
    this.mid = mid
    this.singer = singer
    this.name = name
    this.album = album
    this.duration = duration
    this.image = image
    this.filename = filename
    this.url = url
    this.type = type
    this.lyric = lyric
  }

  public getLyric (): Promise<string> {
    if (this.lyric) {
      return Promise.resolve(this.lyric)
    }
    return new Promise((resolve, reject) => {
      getLyric(this.mid).then(res => {
        const { code, data } = res
        if (code === ERR_OK) {
          this.lyric = Base64.decode(data)
          resolve(this.lyric)
        } else {
          reject(new Error('no lyric'))
        }
      })
    })
  }
}

export function createSong (songData: SongData): Song {
  return new Song({
    id: songData.songid,
    mid: songData.songmid,
    singer: filterSinger(songData.singer),
    name: songData.songname,
    album: songData.albumname,
    duration: songData.interval,
    image: `https://y.gtimg.cn/music/photo_new/T002R300x300M000${songData.albummid}.jpg?max_age=2592000`,
    filename: `C400${songData.songmid}.m4a`,
    url: songData.url,
    type: 'song',
    lyric: ''
  })
}

export function isValid (songData: SongData): boolean {
  return !!(songData.songid && songData.albummid && (!songData.pay || songData.pay.payalbumprice === 0))
}

export function processSongUrl (songs: Song[]): Promise<Song[]> {
  return getSongUrl(songs).then((res: SongUrlMap) => {
    songs = songs.filter(song => {
      const songUrl = res[song.mid]
      if (songUrl) {
        song.url = songUrl.indexOf('http') === -1 ? `http://dl.stream.qqmusic.qq.com/${songUrl}` : songUrl
        return true
      }
      return false
    })
    return songs
  })
}
