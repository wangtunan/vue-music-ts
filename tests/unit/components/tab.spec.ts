import { shallowMount, Wrapper } from '@vue/test-utils'
import { tabList } from '@/assets/js/data'
import Tab from '@/components/tab/index.vue'

describe('tab.vue', () => {
  let wrapper: Wrapper<Vue>
  beforeEach(() => {
    wrapper = shallowMount(Tab, {
      propsData: {
        list: tabList
      },
      stubs: ['router-link']
    })
  })
  it('match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('passed props.list', () => {
    expect(wrapper.props('list').length).toBe(tabList.length)
  })
  it('default activeIndex', () => {
    expect(wrapper.vm.$data.activeIndex).toBe(0)
  })
  it('no passed props.list', () => {
    wrapper = shallowMount(Tab, {
      stubs: ['router-link']
    })
    expect(wrapper.props('list').length).toBe(0)
  })
  it('render success props.list', () => {
    const tabItems = wrapper.findAll('.tab-item')
    expect(tabItems.length).toBe(tabList.length)
  })
  it('change index after click', () => {
    wrapper.findAll('.tab-item').at(1).trigger('click')
    expect(wrapper.vm.$data.activeIndex).toBe(1)
  })
})
