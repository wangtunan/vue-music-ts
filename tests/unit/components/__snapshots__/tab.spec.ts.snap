// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab.vue match snapshot 1`] = `
<div class="m-tab">
  <div class="tab-container">
    <router-link-stub tag="div" to="/recommend" class="tab-item">
      推荐
    </router-link-stub>
    <router-link-stub tag="div" to="/singer" class="tab-item">
      歌手
    </router-link-stub>
    <router-link-stub tag="div" to="/rank" class="tab-item">
      排行
    </router-link-stub>
    <router-link-stub tag="div" to="/search" class="tab-item">
      搜索
    </router-link-stub>
  </div> <span class="tab-bar" style="left: 12.5%;"></span>
</div>
`;
